tls/s2n_client_hello.c

```diff
@@ -299,11 +326,35 @@ int s2n_process_client_hello(struct s2n_connection *conn)
     return S2N_SUCCESS;
 }
 
-int s2n_client_hello_recv(struct s2n_connection *conn)
+static int s2n_client_hello_process_cb_response(struct s2n_connection *conn, int rc)
 {
-    /* Parse client hello */
-    POSIX_GUARD(s2n_parse_client_hello(conn));
+    if (rc >= 0) {
+        if (conn->config->client_hello_cb_mode == S2N_CLIENT_HELLO_CB_BLOCKING) {
+            if (rc) {
+                conn->server_name_used = 1;
+            }    
+            return 0;
+        }
+        if (conn->config->client_hello_cb_mode == S2N_CLIENT_HELLO_CB_NONBLOCKING) {
+            conn->client_hello.callback_async_await = 1;
+            POSIX_BAIL(S2N_ERR_ASYNC_BLOCKED);
+        } 
+    }
+    /* unknown mode | rc < 0*/
+    POSIX_GUARD(s2n_queue_reader_handshake_failure_alert(conn));
+    POSIX_BAIL(S2N_ERR_CANCELLED);
+}
 
+int s2n_client_hello_recv(struct s2n_connection *conn)
+{   
+    S2N_ERROR_IF(conn->client_hello.callback_async_await == 1, S2N_ERR_ASYNC_BLOCKED);
+
+    if (conn->config->client_hello_cb_mode == S2N_CLIENT_HELLO_CB_BLOCKING ||
+            conn->client_hello.callback_async_done != 1)
+    {
+        /* Parse client hello */
+        POSIX_GUARD(s2n_parse_client_hello(conn));
```

Comment:
Should this just be moved into `if (conn->client_hello.parsed == 0) {` code block ?

Reply:
sadly no.  `client_hello.parsed` in not really just that.  it also also used to ensure that callback is invoked only once.

As per the implementation we don't want to invoke the callback after tls1.3 `HELLO_RETRY_MSG` for the second client hello. But we do want to parse the client_hello message again.

Reply:
I see, maybe it's worth renaming `parsing` into `client_hello_cb_called`?

Reply:
Also do we reset `callback_async_done` flag for `HELLO_RETRY_MSG`? Otherwise we won't parse it, right? I might have missed it somewhere.